txnkv/transaction/pessimistic.go

```diff
@@ -150,156 +168,378 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
+		case kvrpcpb.PessimisticLockKeyResultType_LockResultLockedWithConflict:
+			action.ValuesLock.Lock()
+			if action.Values == nil {
+				action.Values = make(map[string]kv.ReturnedValue, 1)
 			}
-			if deadlock := keyErr.Deadlock; deadlock != nil {
-				return errors.WithStack(&tikverr.ErrDeadlock{Deadlock: deadlock})
+			action.Values[string(mutationsPb[0].Key)] = kv.ReturnedValue{
+				Value:                res.Value,
+				Exists:               res.Existence,
+				LockedWithConflictTS: res.LockedWithConflictTs,
 			}
-
-			// Extract lock from key error
-			lock, err1 := txnlock.ExtractLockFromKeyErr(keyErr)
-			if err1 != nil {
-				return err1
+			if res.LockedWithConflictTs > action.MaxLockedWithConflictTS {
+				action.MaxLockedWithConflictTS = res.LockedWithConflictTs
 			}
-			locks = append(locks, lock)
-		}
-		// Because we already waited on tikv, no need to Backoff here.
-		// tikv default will wait 3s(also the maximum wait value) when lock error occurs
-		if resolvingRecordToken == nil {
-			token := c.store.GetLockResolver().RecordResolvingLocks(locks, c.startTS)
-			resolvingRecordToken = &token
-			defer c.store.GetLockResolver().ResolveLocksDone(c.startTS, *resolvingRecordToken)
-		} else {
-			c.store.GetLockResolver().UpdateResolvingLocks(locks, c.startTS, *resolvingRecordToken)
-		}
-		resolveLockOpts := txnlock.ResolveLocksOptions{
-			CallerStartTS: 0,
-			Locks:         locks,
+			action.ValuesLock.Unlock()
+		case kvrpcpb.PessimisticLockKeyResultType_LockResultFailed:
+			failedMutations.AppendMutation(batch.mutations.GetMutation(0))
+		default:
+			panic("unreachable")
 		}
+	}
+
+	if len(lockResp.Results) > 0 && failedMutations.Len() < len(lockResp.Results) {
+		// Can we do it like this?
+
 		if action.LockCtx.Stats != nil {
-			resolveLockOpts.Detail = &action.LockCtx.Stats.ResolveLock
+			action.LockCtx.Stats.MergeReqDetails(diagCtx.reqDuration, batch.region.GetID(), diagCtx.sender.GetStoreAddr(), lockResp.ExecDetailsV2)
 		}
-		resolveLockRes, err := c.store.GetLockResolver().ResolveLocksWithOpts(bo, resolveLockOpts)
+	}
+
+	// The primary must be locked first, otherwise there should be some bug in the implementation.
+	if batch.isPrimary && failedMutations.Len() < len(lockResp.Results) && lockResp.Results[0].Type == kvrpcpb.PessimisticLockKeyResultType_LockResultFailed {
+		return true, nil, errors.New("Pessimistic lock response corrupted")
+	}
+
+	var locks []*txnlock.Lock
+	for _, keyErr := range keyErrs {
+		// Check already exists error
+		if alreadyExist := keyErr.GetAlreadyExist(); alreadyExist != nil {
+			e := &tikverr.ErrKeyExist{AlreadyExist: alreadyExist}
+			return true, nil, c.extractKeyExistsErr(e)
+		}
+		if deadlock := keyErr.Deadlock; deadlock != nil {
+			return true, nil, errors.WithStack(&tikverr.ErrDeadlock{Deadlock: deadlock})
+		}
+
+		// Extract lock from key error
+		lock, err1 := txnlock.ExtractLockFromKeyErr(keyErr)
+		if err1 != nil {
+			return true, nil, err1
```

Comment:
After the key was forcibly locked in tikv, would it still be a `WriteConflictError` returned here to the upper layer to inform the statement must be retried?